pkg/yurthub/multiplexer/manager.go

```diff
@@ -0,0 +1,178 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multiplexer
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/generic/registry"
+	kstorage "k8s.io/apiserver/pkg/storage"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	ystorage "github.com/openyurtio/openyurt/pkg/yurthub/multiplexer/storage"
+)
+
+var NamespaceKeyFunc = func(obj runtime.Object) (string, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return "", err
+	}
+
+	return registry.NamespaceKeyFunc(request.WithNamespace(request.NewContext(), accessor.GetNamespace()), "", accessor.GetName())
+}
+
+var NoNamespaceKeyFunc = func(obj runtime.Object) (string, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return "", err
+	}
+
+	return registry.NoNamespaceKeyFunc(request.WithNamespace(request.NewContext(), accessor.GetNamespace()), "", accessor.GetName())
+}
+
+type MultiplexerManager interface {
+	ResourceCacheConfig(gvr *schema.GroupVersionResource) (*ResourceCacheConfig, error)
+	ResourceCache(gvr *schema.GroupVersionResource) (Interface, func(), error)
+}
+
+type multiplexerManager struct {
+	restStoreManager    ystorage.StorageManager
+	discoveryClient     discovery.DiscoveryInterface
+	cacheMap            map[string]Interface
+	cacheConfigMap      map[string]*ResourceCacheConfig
+	cacheDestroyFuncMap map[string]func()
+}
+
+func NewShareCacheManager(
+	restStoreManager ystorage.StorageManager,
+	discoveryClient discovery.DiscoveryInterface) MultiplexerManager {
+	return &multiplexerManager{
+		restStoreManager:    restStoreManager,
+		discoveryClient:     discoveryClient,
+		cacheMap:            make(map[string]Interface),
+		cacheConfigMap:      make(map[string]*ResourceCacheConfig),
+		cacheDestroyFuncMap: make(map[string]func()),
+	}
+}
+
+func (scm *multiplexerManager) ResourceCacheConfig(gvr *schema.GroupVersionResource) (*ResourceCacheConfig, error) {
+	if config, ok := scm.cacheConfigMap[gvr.String()]; ok {
+		return config, nil
+	}
+
+	apiResource, err := scm.getAPIResource(gvr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get api resource from %s", gvr.String())
+	}
+
+	gvk, listGVK := scm.convertToGVK(gvr, apiResource)
+
+	config := scm.newResourceCacheConfig(gvk, listGVK, apiResource)
+
+	scm.cacheConfigMap[gvr.String()] = config
+	return config, nil
+}
+
+func (scm *multiplexerManager) getAPIResource(gvr *schema.GroupVersionResource) (metav1.APIResource, error) {
+	resourceList, err := scm.discoveryClient.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
```

Comment:
Is it possible not make an `APIResource` request to kube-apiserver? I'm afraid that multiplexer cache maybe effected when cloud-edge network offline.

Reply:
Okay, I’ve adjusted this part of the implementation to use DefaultRestMapper for converting GVR to GVK.